Added documentation about cross shard transactions (#8014)
near/nearcore: docs/architecture/how/cross-shard.md

# Cross shard transactions - deep dive

In this article, we'll look deeper into how cross-shard transactions are working on the simple example of user 'shard0' transfering money to user 'shard1'.

These users are on separate shards (shard0 is on shard 0 and shard1 is on shard 1).

Imagine, we call the following command line:

```shell
NEAR_ENV=local near send shard0 shard1 500 
```

What happens under the hood? How is this transaction changed into receipts and processed by near?

## From Explorer perspective
If you look at simple token transfer in explorer (for example: https://explorer.testnet.near.org/transactions/6cNJpNKWP55YrxrgRzc2gi6BM91fo3V6n2vVAo41MvZv), you can see that it is broken into three separate sections:

* convert transaction into receipt ( executed in block B )
* receipt that transfers tokens ( executed in block B+1 )
* receipt that refunds gas ( executed in block B+2 )

But under the hood, the situation is a little bit more complex, as there is actually one more receipt (that is created after converting the transaction). Let's take a deeper look.

## Internal perspective (Transactions & Receipts)

> One important think to remember, is that NEAR is sharded - so in all our designs, we have to assume that each account is on a separate shard (so that the fact that some of them are colocated doesn't give any advantage).

Reviewer: ```suggestion
One important thing to remember, is that NEAR is sharded - so in all our designs, we have to assume that each account is on a separate shard. So that the fact that some of them are colocated doesn't give any advantage.
```

### Step 1 - Transaction
This is the part that we receive from the user (SignedTransaction) - it has 3 parts:

* signer (account + key) who signed the transaction
* receiver (in which account context should we execute this)
* payload - a.k.a Actions to execute.

As the first step, we want to change this transaction into a Receipt (a.k.a 'internal' message) - but before doing that, we must verify that:

* the message signature matches (that is - that this message was actually signed by this key)
* that this key is authorized to act on behalf that account (so it is a full access key to this account - or a valid fuction key).

The last point above means, that we MUST execute this Transaction to Receipt transition within the shard that the ``signer`` belongs to (as other shards don't know the state that belongs to signer - so they don't know which keys it has).

So actually if we look inside the chunk 0 (where shard0 belongs to) at block B, we'll see the transaction:

> ```json

Reviewer 2: this doesn't look like json

Author: removed.

            ),
            transactions: [
                SignedTransaction {
                    transaction: Transaction {
                        signer_id: AccountId(
                            "shard0",
                        ),
                        public_key: ed25519:Ht8EqXGUnY8B8x7YvARE1LRMEpragRinqA6wy5xSyfj5,
                        nonce: 11,
                        receiver_id: AccountId(
                            "shard1",
                        ),
                        block_hash: 6d5L1Vru2c4Cwzmbskm23WoUP4PKFxBHSP9AKNHbfwps,
                        actions: [
                            Transfer(
                                TransferAction {
                                    deposit: 500000000000000000000000000,
                                },
                            ),
                        ],
                    },
                    signature: ed25519:63ssFeMyS2N1khzNFyDqiwSELFaUqMFtAkRwwwUgrPbd1DU5tYKxz9YL2sg1NiSjaA71aG8xSB7aLy5VdwgpvfjR,
                    hash: 6NSJFsTTEQB4EKNKoCmvB1nLuQy4wgSKD51rfXhmgjLm,
                    size: 114,
                },
            ],
            receipts: [],
        },
    ),
)
```

```

> Side note: When we're changing the transaction into receipt, we also use this moment to deduct prepaid gas fees and transfered tokens from the 'signer' account.

Reviewer: > and transfered tokens

This deserves more than a side note.

Author: yep - we'll create a separate article about how much gas we charge etc.

                    ),
                    receiver_id: AccountId(
                        "shard1",
                    ),
                    receipt_id: 3EtEcg7QSc2CYzuv67i9xyZTyxBD3Dvx6X5yf2QgH83g,
                    receipt: Action(
                        ActionReceipt {
                            signer_id: AccountId(
                                "shard0",
                            ),
                            signer_public_key: ed25519:Ht8EqXGUnY8B8x7YvARE1LRMEpragRinqA6wy5xSyfj5,
                            gas_price: 103000000,
                            output_data_receivers: [],
                            input_data_ids: [],
                            actions: [
                                Transfer(
                                    TransferAction {
                                        deposit: 500000000000000000000000000,
                                    },
                                ),
                            ],
                        },
                    ),
                },
            ],
        },
    ),
)
```

```
Side comment: notice that receipt itself no longer has a ``signer`` field, but a ``predecessor_id`` one.
```

> Such receipt is magically (we'll explain this magic in a separate article) sent to the destination shard, where it can be executed.

Reviewer: Personal preference to avoid magic in designs and documentation.

Author: ok :-(

## 3. Gas refund.

> When shard 1 processes the receipt above, it is then ready to refunt the unused gas to the original account (shard0). So it also creates the receipt, and puts it inside the chunk. This time it is in shard1 (as that's where it was executed).

Reviewer: ```suggestion
When shard 1 processes the receipt above, it is then ready to refund the unused gas to the original account (shard0). So it also creates the receipt, and puts it inside the chunk. This time it is in shard1 (as that's where it was executed).
```

> When shard 1 processes the receipt above, it is then ready to refunt the unused gas to the original account (shard0). So it also creates the receipt, and puts it inside the chunk. This time it is in shard1 (as that's where it was executed).

Reviewer: This sentence has both "shard 1" and "shard1". Should we use formatting to separate account names from "shard 1". Something like `shard1` would work.

Author: done.

            receipts: [
                Receipt {
                    predecessor_id: AccountId(
                        "system",
                    ),
                    receiver_id: AccountId(
                        "shard0",
                    ),
                    receipt_id: 6eei79WLYHGfv5RTaee4kCmzFx79fKsX71vzeMjCe6rL,
                    receipt: Action(
                        ActionReceipt {
                            signer_id: AccountId(
                                "shard0",
                            ),
                            signer_public_key: ed25519:Ht8EqXGUnY8B8x7YvARE1LRMEpragRinqA6wy5xSyfj5,
                            gas_price: 0,
                            output_data_receivers: [],
                            input_data_ids: [],
                            actions: [
                                Transfer(
                                    TransferAction {
                                        deposit: 669547687500000000,
                                    },
                                ),
                            ],
                        },
                    ),
                },
            ],
        },
    ),
)
```

> Such gas refunds receipts are a little bit special - as we'll set the predecessor_id to be ``system`` - but the receiver is what we expect (shard0 account).

Reviewer: I remember having a question, whether `system` belongs to any particular shard.

So putting it all together would look like this:

![image](https://user-images.githubusercontent.com/91919554/200617392-00b9fa0c-2f15-40ad-9802-137ca9a5a15d.png)

> But wait - NEAR was saying that transfers are happening with 2 blocks - but I here I see that it took 3 blocks. What's wrong?

Reviewer: ```suggestion
But wait - NEAR was saying that transfers are happening with 2 blocks - but here I see that it took 3 blocks. What's wrong?
```

The image above is a simplification, and reality is a little bit tricker - especially as receipts in a given chunks are actually receipts received as a result from running a PREVIOUS chunk from this shard.

We'll explain it more in the next section.

# Advanced: What's actually going on?

As you could have read in [tx_receipts.md] - the 'receipts' field in chunk is actually representing 'outgoing' receipts from previous block.

So our image should look more like this:
![image](https://user-images.githubusercontent.com/91919554/200621066-a5d06f2d-ff43-44ce-a52b-47dc44d6f8ab.png)

In this example, the black boxes are representing the 'processing' of the chunk, and red arrows are cross-shard communication.

> So when we process Shard 0 from block 1676, we read the transation, and output the receipt - which later magically travels to become the input for shard1 in block 1677.

Reviewer: ```suggestion
So when we process Shard 0 from block 1676, we read the transation, and output the receipt - which later becomes the input for shard1 in block 1677.
```
